src/typechecker/FStar.TypeChecker.Util.fs

```diff
@@ -1011,26 +1011,96 @@ let check_trivial_precondition env c =
   
   ct, vc, Env.guard_of_guard_formula <| NonTrivial vc
 
-let maybe_coerce_bool_to_type env (e:term) (lc:lcomp) (t:term) : term * lcomp =
-    if env.is_pattern then e, lc else
+let coerce_with (env:Env.env) (e : term) (lc : lcomp) (f : lident) (ty : typ) : term * lcomp =
+    match Env.try_lookup_lid env f with
+    | Some _ ->
+        if Env.debug env (Options.Other "Coercions") then
+            BU.print1 "Coercing with %s!\n" (Ident.string_of_lid f);
+        let coertion = S.fvar (Ident.set_lid_range f e.pos) (Delta_constant_at_level 1) None in
+        let lc = bind e.pos env (Some e) lc (None, U.lcomp_of_comp <| S.mk_Total ty) in
```

Comment:
Why do we need to do this `bind`, given that coercions are inserted in the first phase?

Reply:
I think (I wrote this long ago) this was an attempt to keep the WPs well-typed, since `set_result_typ_lc` will not guarantee that, right? But yeah, since it's phase 1, it doesn't really matter. Doing that works, so if you think it's preferable I can do that.

Reply:
This is fine, `bind` itself does some shortcuts in the `lax` mode. Thanks.

Reply:
I guess the discussion about doing coercion insertion regardless of phases impacts this change. 

If we insert coercions on both phases, then this bind should remain … ?

Reply:
If we are in the 2-phases mode, then the coercion would have been inserted in the first phase, and the second phase will not insert it again, so this part of code should not get exercised in the second phase.